### ISSUE TITLE
BE - removeField does not remove field

### ISSUE THREAD

**Issue by kamilmielnik** [metabase/metabase#36575]
Found alongside #36574. I tried to fix #36574 by removing fields manually on frontend side, but `removeField` did not work.
It's not high priority as it's currently not blocking frontend but is seems there's an issue nonetheless.

```tsx
console.log(Lib.fields(query, 1).length); // 3

const finalQuery = Lib.fields(query, 1).reduce((nextQuery, field) => {
  return Lib.removeField(nextQuery, 1, field);
}, query);

console.log(Lib.fields(finalQuery, 1).length); // 3
```

### How to reproduce

1. `git checkout 35982-expression-step`
2. apply this patch
```patch
diff --git a/frontend/src/metabase/query_builder/components/notebook/steps/ExpressionStep.tsx b/frontend/src/metabase/query_builder/components/notebook/steps/ExpressionStep.tsx
index 8f9bd5fd92..ff1d92b397 100644
--- a/frontend/src/metabase/query_builder/components/notebook/steps/ExpressionStep.tsx
+++ b/frontend/src/metabase/query_builder/components/notebook/steps/ExpressionStep.tsx
@@ -70,10 +70,30 @@ export const ExpressionStep = ({
       )}
       isLastOpened={isLastOpened}
       onRemove={clause => {
+        console.log("A", Lib.fields(query, 1).length); // 4 - correct
+
+        const finalQuery1 = Lib.fields(query, 1).reduce((nextQuery, field) => {
+          return Lib.removeField(nextQuery, 1, field);
+        }, query);
+
+        console.log("B", Lib.fields(finalQuery1, 1).length); // 4 - wrong, expected: 0
+
         const nextQuery = Lib.dropStageIfEmpty(
           Lib.removeClause(query, stageIndex, clause),
           step.stageIndex,
         );
+
+        console.log("C", Lib.fields(nextQuery, 1).length); // 3 - correct
+
+        const finalQuery2 = Lib.fields(nextQuery, 1).reduce(
+          (nextQuery, field) => {
+            return Lib.removeField(nextQuery, 1, field);
+          },
+          nextQuery,
+        );
+
+        console.log("D", Lib.fields(finalQuery2, 1).length); // 3 - wrong, expected 0
+
         updateQuery(nextQuery);
       }}
       withLegacyPopover
```
3. run `19745-cc-nested-query-remove-expressions.cy.spec.js` e2e test
4. observe JS console

**Comment by metamben** [comment]
@kamilmielnik there is some confusion here. I see `removeField` has been declared to return `ColumnMetadata[]`, but in fact it returns opaque references. `removeField` won't find those references and won't remove anything. I'll need to understand better what you need to achieve what you want.

I find the e2e test confusing too. How can one create such a query (i.e. with `fields` set after the summary) in the UI?

**Comment by kamilmielnik** [comment]
> I'll need to understand better what you need to achieve what you want.

I was trying to "fix" #36574 by removing fields manually on the frontend side.
Perhaps this does not make any sense at all.

> I find the e2e test confusing too. How can one create such a query (i.e. with `fields` set after the summary) in the UI?

https://github.com/metabase/metabase/assets/6830683/f9ce47a4-a6eb-4d39-ab10-7c414ec38d37

**Comment by metamben** [comment]
@kamilmielnik thanks for the explanation. you could do something similar by removing `returnedColumns`, but I guess it's better to fix #36574 properly.

**Comment by metamben** [comment]
We decided this is not a bug and the main problem will be fixed for #36574.
